8242548: Wrapped labeled controls using -fx-line-spacing cut text off

hjohn opened the pull request:
This is a solution for 8242548.  There was zero test coverage, so I added a few tests for this as well.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8242548](https://bugs.openjdk.java.net/browse/JDK-8242548): Wrapped labeled controls using -fx-line-spacing cut text off


### Reviewers
 * Ajit Ghaisas ([aghaisas](@aghaisas) - **Reviewer**)
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/173/head:pull/173`
`$ git checkout pull/173`

kevinrushforth commented:
@aghaisas can you review this?

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
@aghaisas can you also review this?

aghaisas commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/skin/Utils.java:
```diff
@@ -417,12 +417,14 @@ public static String computeClippedText(Font font, String text, double width,
             throw new IllegalArgumentException("Must specify a font");
         }
 
+        height += lineSpacing;
```
Modifying 'height' parameter seems incorrect as this parameter is used in other calculations below. If needed, use a separate local variable to use 'height+lineSpacing'

hjohn replied:
The height used needs to be changed everywhere.  However, I can move this to the caller instead, with a comment explaining why:

> The computeClippedWrappedText call works with full lines (height of line + line spacing), including the last line.  However the wrap height does not include the line spacing for the last line.  In order to avoid wrapping the last line of text even though there is sufficient space the wrap height passed to computeClippedWrappedText is increased with the line spacing so it computes the correct text.

aghaisas replied:
I see your point. I am OK with the changes that you have done.

aghaisas commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/skin/Utils.java:
```diff
@@ -417,12 +417,14 @@ public static String computeClippedText(Font font, String text, double width,
         }
 
+        height += lineSpacing;
+
```
This method makes a call to -
double eHeight = computeTextHeight(font, ellipsis, 0, boundsType);

What is the behavior if we make a call to the other method that takes in lineSpacing?
double eHeight = computeTextHeight(font, ellipsis, 0, lineSpacing, boundsType);

hjohn replied:
The call only determines the ellipsis height (eHeight) to see if the space is so small that even an ellipsis wouldn't fit.  For a multi-line ellipsis (if there is such a thing) we could pass in the line spacing here.

aghaisas commented:
I tested with the test program in the bug and can confirm that this PR fixes the reported issue.

I found that the unit test (together with stub) passes without the code fix as well. 
I think, this is a limitation of our test framework that we need to do code fix + stub fix as well. I find that you have done these changes.

hjohn commented:
I added a comment clarification and added passed the line spacing to the ellipsis height calculation as well (just in case there is a multi line ellipsis). 

The test indeed doesn't quite cover the change, the StubTextLayout is not advanced enough to actually do line wrapping calculations (it ignores wrap height), and would need to be updated.

kevinrushforth commented:
@hjohn Updates to a PR under review should be done as additional commits. Please do not force push to your branch. This makes it more difficult for reviewers, since the incremental diffs from the last time it was reviewed are lost. A force push of your branch should only be used in extraordinary circumstances.

aghaisas commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/skin/Utils.java:
```diff
@@ -417,17 +417,27 @@ public static String computeClippedText(Font font, String text, double width,
+        // However the calculations include the line spacing as part of a
+        // line's height.  In order to not cut off the last line because its
+        // line spacing wouldn't fit, the height used for the calculation
+        // is increase here with the line spacing amount.
```
increase -> increased

kevinrushforth commented:
The code change and test look fine. Please change the PR title to match the JBS issue title, so:

```
8242548: Wrapped labeled controls using -fx-line-spacing cut text off
```

aghaisas commented:
@hjohn, this PR is ready to be merged.
You need to comment /integrate on this PR as instructed by the bot above. I will sponsor it once you do it.

hjohn commented:
/integrate

hjohn commented:
Sorry I figured only a committer was allowed to use that command.

aghaisas commented:
/sponsor
